### DISCUSSION TITLE
Storage tradeoffs | public bucket vs signedURL

### DISCUSSION THREAD

**Discussion by zidkim** [supabase/supabase#6458] [Questions] [answered]
What are the pros and cons between using a public bucket and a signedURL with an expiry date of 10 years?
The goal is keep the images stored and accessible from db forever or as long as possible.

By utilizing a public bucket, any user, even those from outside the app can run a cascade of infinite loops to download whatever image is stored in the public bucket. Advantage here of course is the non-expiry date of the download url.

For storage, is there DDOS protection, rate limiting per IP, etc? 
With respect to cost, even though these images are stored in cache, I imagine each cache hit incurs a download cost?

Now, for the private bucket + signedURL approach -- we can set rules so that users from outside the app can't download the image. Plus, the 10 year expiry date is quite long, so that's good. 

And what other factors should I be considering?

**Comment by GaryAustin1** [comment, marked as answer]
I'm not sure what advantage you get with a long term signedURL approach.  If someone gets the url (just like in the public case) then they can use it the same way.  RLS protections only applied when you asked for the signedURL not once you start using it.  

As far as limits, it is running thru the CDN and I notice limit headers on the files, but I have no idea on the details.

**Reply by zidkim** [reply]
With private buckets (requiring auth), you would have to be inside the application to access the bucket.
To reproduce, simply try to access an asset from private bucket in incognito.
This provides at least 1 layer of protection.

With this in mind, I'm not sure what the benefit of using public buckets are, aside from the peace of mind due to the no-expiry-date urls. That being said, if there were security measures already embedded into Supabase, such as rate limiting per IP and DDos attacks, then I believe this would not be an issue. Mind you, my security knowledge in general is beginner at best. It just seems easy for an attacker to run infinite loop requests to multiple assets, across multiple IPs, racking up $$$.

**Reply by GaryAustin1** [reply]
If you provide a signed URL anyone who has it can use it.  They don't have to be a user of the app.
The main difference in a public bucket is the URL does not need the token attached.  You can't access a public bucket as it RLS locked down other than with the URL.  The urls go straight to S3 and don't touch the database.

But I'm not discouraging you from using private buckets, just pointing out your URL with a long lived token is no different than a public URL.
https://github.com/supabase/supabase/discussions/3937

**Reply by zidkim** [reply]
Ah, I forgot to delineate that its the token that is the difference maker here.
To your point, I see what youre saying, thank you for shedding more light on how RLS works.
There are still many question marks around current and future security features.
For this I'll reach out to the team directly, and hopefully there's some good news there.
Appreciate the help/insights @GaryAustin1

**Reply by pfcodes** [reply]
@GaryAustin1 how does caching work? are items in private buckets cached? i'd assume that the token querystring might cause some issues with caching.
